sound/soc/sof/intel/mtl.c

```diff
@@ -612,6 +613,54 @@ static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 	return mtl_disable_interrupts(sdev);
 }
 
+static snd_pcm_uframes_t mtl_dsp_pcm_pointer(struct snd_sof_dev *sdev,
+					     struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_component *scomp = sdev->component;
+	struct hdac_stream *hstream = substream->runtime->private_data;
+	struct sof_ipc_stream_posn *posn;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t pos;
+	u32 host_l, host_u;
+	u32 llp_l, llp_u;
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	if (!spcm) {
+		dev_warn_ratelimited(sdev->dev, "warn: can't find PCM with DAI ID %d\n",
+				     rtd->dai_link->id);
+		return 0;
+	}
+
+	posn = &spcm->stream[substream->stream].posn;
+
+	pos = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, AZX_REG_VS_SDXDPIB_XBASE +
+			       (AZX_REG_VS_SDXDPIB_XINTERVAL * hstream->index));
+
+	if (pos >= hstream->bufsize)
+		pos %= hstream->bufsize;
+
+	pos = bytes_to_frames(substream->runtime, pos);
+
+	posn->wallclock = snd_sof_dsp_read64(sdev, HDA_DSP_BAR, MTL_HFDSPWC);
+
+	host_l = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPHCLLPL_BASE +
+				 MTL_PPHCLLP_INTERVAL * hstream->index);
+	host_u = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPHCLLPU_BASE +
+				 MTL_PPHCLLP_INTERVAL * hstream->index);
+	posn->host_posn = bytes_to_frames(substream->runtime, PACK_TWO_U32_TO_U64(host_l, host_u));
+
+	llp_l = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPL_BASE +
+				 MTL_PPHCLLP_INTERVAL * hstream->index);
+	llp_u = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPU_BASE +
+				 MTL_PPHCLLP_INTERVAL * hstream->index);
+	posn->dai_posn = PACK_TWO_U32_TO_U64(llp_l, llp_u);
+
+	trace_sof_intel_hda_dsp_pcm(sdev, hstream, substream, pos);
+
+	return pos;
```

Comment:
No, this is not right. We've always said that the .pointer callback returns the DPIB value. It's the host DMA position.

Reply:
pos is the value from DPIB.

	pos = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, AZX_REG_VS_SDXDPIB_XBASE +
			       (AZX_REG_VS_SDXDPIB_XINTERVAL * hstream->index));

Reply:
this function also calculate "struct sof_ipc_stream_posn *posn" for dai position